{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2cc88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "clinical_df = pd.read_excel(\"../input/1-s2.0-S0092867420301070-mmc1.xlsx\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a72c90b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_dfs = pd.read_excel('../input/1-s2.0-S0092867420301070-mmc2.xlsx', sheet_name= None)\n",
    "\n",
    "global_df = proteomics_dfs['A-global-proteomics']\n",
    "phospho_df = proteomics_dfs['B-phospho-proteomics']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d5ffad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of cases: 153\n",
      "Tumor cases: 95\n",
      "Serous cases: 12\n",
      "Endometrioid cases: 83\n"
     ]
    }
   ],
   "source": [
    "# First, read the table of case numbers, and filter out the ones that are not excluded \n",
    "# Filter out the excluded case\n",
    "print(\"Total number of cases:\", len(clinical_df))\n",
    "case_df = clinical_df[clinical_df['Case_excluded'] == 'No']\n",
    "case_df = case_df[case_df['Histologic_type'].isin(['Endometrioid','Serous'])]\n",
    "print(\"Tumor cases:\", len(case_df))\n",
    "\n",
    "serous_df = case_df[case_df['Histologic_type'] == 'Serous']\n",
    "endometrioid_df = case_df[case_df['Histologic_type'] == 'Endometrioid']\n",
    "print(\"Serous cases:\", len(serous_df))\n",
    "print(\"Endometrioid cases:\", len(endometrioid_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bcb409a",
   "metadata": {},
   "source": [
    "Q2-easy: What is the average age of patients with serous tumor samples analyzed in the study?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ed1edbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average age of serous cases: 68.5\n"
     ]
    }
   ],
   "source": [
    "#serous_df['Age'].to_list()\n",
    "# average age of serous cases\n",
    "serous_age = serous_df['Age'].mean()\n",
    "print(\"Average age of serous cases:\", serous_age)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9c07f2f",
   "metadata": {},
   "source": [
    "Q1-hard: What is the correlation between the abundance of the protein PLK1 and the abundance of CHEK2-S163 in tumor samples? Exclude samples not in the study and with missing values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a7bf6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spearman correlation coefficient: 0.4764813684787027\n",
      "P-value: 0.0004084391725761019\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import spearmanr\n",
    "import numpy as np\n",
    "\n",
    "# Filter the global proteomics data for tumor cases \n",
    "case_ids = case_df['idx'].tolist()\n",
    "global_df = global_df.filter(items=case_ids+[\"idx\"], axis=1)\n",
    "phospho_df = phospho_df.filter(items=case_ids+[\"idx\"], axis=1)\n",
    "\n",
    "plk1 = global_df[global_df['idx'] == 'PLK1'].values.tolist()[0][:-1]\n",
    "chek2 = phospho_df[phospho_df['idx'] == 'CHEK2-S163'].values.tolist()[0][:-1]\n",
    "\n",
    "plk1 = np.asarray(plk1)[~np.isnan(chek2)]\n",
    "chek2 = np.asarray(chek2)[~np.isnan(chek2)]\n",
    "\n",
    "# Calculate the Spearman correlation\n",
    "spearman_corr, p_value = spearmanr(plk1, chek2)\n",
    "print(\"Spearman correlation coefficient:\", spearman_corr)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b52813e",
   "metadata": {},
   "source": [
    "Q3-hard: What is the age of the patient associated with the lowest APM-Z score?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f99db295",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18    60.0\n",
       "Name: Age, dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmb_df = pd.read_excel('../input/1-s2.0-S0092867420301070-mmc7.xlsx', sheet_name=[\"B-APM subtypes\"])['B-APM subtypes']\n",
    "# find the idx where APP_Z_score is max\n",
    "min_idx = tmb_df.loc[tmb_df['APP_Z_score'].idxmin()]['idx']\n",
    "\n",
    "# find the idx where APP_Z_score is max\n",
    "case_df[case_df['idx']==min_idx]['Age']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af580b10",
   "metadata": {},
   "source": [
    "Q5-hard: What is the median number of variants per Mbp for the serous tumor samples in the study? (Note: the variable is expressed in log2 in the input table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f163fba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/eylai/Projects/LLMBenchmark/.venv/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:3860: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "/Users/eylai/Projects/LLMBenchmark/.venv/lib/python3.10/site-packages/numpy/_core/_methods.py:145: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "np.float64(nan)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "tmb_df = tmb_df[tmb_df['idx'].isin(serous_cases)]\n",
    "vpm = tmb_df['Log2_variant_per_Mbp'].values\n",
    "vpm = 2**vpm\n",
    "np.median(vpm)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d46a9b",
   "metadata": {},
   "source": [
    "Q6-easy: What the most common FIGO stage of the samples taken amongst patients 70 or above?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672acf28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FIGO_stage\n",
       "IA       12\n",
       "IB        4\n",
       "IIIA      2\n",
       "IVB       2\n",
       "II        2\n",
       "IIIC1     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "above_70 = case_df[case_df['Age'] >= 70]\n",
    "above_70['FIGO_stage'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf99e39",
   "metadata": {},
   "source": [
    "Q7-hard: How many are the significant genes by acetylproteomics? (parse value from spreadsheet README)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e7fa84b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BRD8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DHX15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FUS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PARP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRIM33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     BRD8\n",
       "0   DHX15\n",
       "1     SSB\n",
       "2     FUS\n",
       "3   PARP1\n",
       "4  TRIM33"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in acetyl sheet: 16\n"
     ]
    }
   ],
   "source": [
    "dfs = pd.read_excel('../input/1-s2.0-S0092867420301070-mmc3.xlsx', sheet_name=None)\n",
    "gene_readme = dfs['README']\n",
    "\n",
    "acetyl_sheet = gene_readme.loc[gene_readme[\"Description\"].str.contains(\"acetylproteomics\", case=False), \"Sheet\"].to_list()[0]\n",
    "acetyl_sheet\n",
    "\n",
    "# Since the sheet is technically a list, the first item gets considered a header and we have to add 1\n",
    "num_genes = len(dfs[acetyl_sheet])+1\n",
    "display(dfs[acetyl_sheet].head()) \n",
    "print(\"Number of genes in acetyl sheet:\", num_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56303b11",
   "metadata": {},
   "source": [
    "Hard question: Which proteins found in both comparisons are targeted by FDA-approved drugs?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e8a7ee2",
   "metadata": {},
   "source": [
    "Q9-easy: What is the difference between the average false discovery rate (FDR) in CBX3 genes and the average FDR for the rest of the genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69eca2f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean FDR for CBX3: 0.016840668972299792\n",
      "Mean FDR for other genes: 0.02484464434616611\n",
      "Difference in mean FDR: -0.008003975373866316\n"
     ]
    }
   ],
   "source": [
    "fdr_df = dfs['F-SS-phospho']\n",
    "fdr_df[['Gene','FDR.phos']]\n",
    "\n",
    "# fdr_df.groupby('FDR.phos').mean()\n",
    "gene_mean = fdr_df.groupby('Gene')['FDR.phos'].mean()\n",
    "cbx3 = gene_mean['CBX3']\n",
    "\n",
    "other_genes = fdr_df[fdr_df['Gene'] != 'CBX3']\n",
    "other = np.mean(other_genes['FDR.phos'])\n",
    "\n",
    "print(\"Mean FDR for CBX3:\", cbx3)\n",
    "print(\"Mean FDR for other genes:\", other)\n",
    "print(\"Difference in mean FDR:\", cbx3 - other)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00bca39e",
   "metadata": {},
   "source": [
    "Hard question: "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
